### ISSUE TITLE
Indiana Jones and the Infernal Machine

### ISSUE THREAD

**Issue by IllustriousProsperus** [narzoul/DDrawCompat#382]
First of all loading the game is difficult, there is a **black rectangle**, if there are saves and you click on the black background and if you try to move a save, the black screen disappears. After loading the game, if you want to change the game options, some screens are not visible but in low res. If you manage to open a screen setting, if you press **"ESC" there is a flash (1 sec.)** before the game can start working again. Maybe there is something wrong with how the game forces the use of **win98 style windows**. 
**_The **"FullscreenMode = borderless"** to **"FullscreenMode = exclusive"** crashes the game._**

black rectangle (without "ddrawcompact" the rectangle is white)
![Screenshot 2024-10-25 134234](https://github.com/user-attachments/assets/331911e1-80a1-43ab-be16-96c26ea67f23)
after a "move and drag"
![Screenshot 2024-10-25 134244](https://github.com/user-attachments/assets/e0098ba6-cf79-4cb0-b13a-fa6d1b04f268)
saving/loading is a little complicated
![Screenshot (66)](https://github.com/user-attachments/assets/9d62f5f2-4372-412c-ad58-6cfb46959729)
Working windows (?) you don't see the pointer in the screenshot but it is present. And it's stuck from a normal pointer to a loading pointer.
![Screenshot (67)](https://github.com/user-attachments/assets/50069e04-35db-45a5-977e-9e520bcb6bae)

_**"Comparison with DDrawCompat-v0.5.4 and DgVoodoo2 2.82.3"**_

**_DDrawCompat-v0.5.4_**
**_Exiting the game (note the exit image and lowres)_**
![Screenshot (68)](https://github.com/user-attachments/assets/da111204-5834-46f2-9d4f-9e5cbd56e3e9)

**_Game Options (lowres)_**
![Screenshot (69)](https://github.com/user-attachments/assets/af1eb792-9774-48d6-a7c8-5159f8de73fb)

**_savegame (lowres and bugged)_**
![Screenshot (73)](https://github.com/user-attachments/assets/d064759c-ab01-4d35-931c-f8501b9fa1c2)

_**"Dgvoodoo2 2.82.3"**_

**_(no exit image wrong position but hires)_**
![Screenshot (70)](https://github.com/user-attachments/assets/ff60ce7e-e8da-427f-8844-a312de6ada39)

**_(note the style of the window and wrong position but hires)_**
![Screenshot (71)](https://github.com/user-attachments/assets/f1b26b15-feb4-4cc7-890a-a28dbce90982)

**_(highres and bugged)_**
![Screenshot (72)](https://github.com/user-attachments/assets/eda2c792-e70e-4963-b6fb-3e049cc2ea61)

**Comment by BEENNath58** [comment]
The black window should be from the usage of the "square styled" Windows NT classic type window used in DDrawCompat if I am not wrong.

**Comment by narzoul** [comment]
See comment here: https://github.com/narzoul/DDrawCompat/issues/410#issuecomment-2526213356

I didn't really understand what you meant by "lowres", it looks like the proper resolution to me.

**Comment by Spatzendame** [comment]
> See comment here: [#410 (comment)](https://github.com/narzoul/DDrawCompat/issues/410#issuecomment-2526213356)
> 
> I didn't really understand what you meant by "lowres", it looks like the proper resolution to me.

Thanks for resolving the file list. I will try the hex changes.
I think with "lowres" he means the small window boxes. lowres = small. This is a GDI scaling issue. Can be fixed in compatibility mode to prevent GDI scaling by System (Advanced).

**Comment by Spatzendame** [comment]
I tried your recommended hex edits and it really resolved the issue with the loading screen. But it caused another new issue with the trading screen at the end of a level. Watch the screenshot.
![indy3D_narzoul_ddraw2](https://github.com/user-attachments/assets/9e794b05-49fe-45cc-b9d4-55827008e612)

Let's find a solution for this too and then everything is okay.

**Comment by narzoul** [comment]
Do you have a save game to quickly reproduce the issue? I don't see how this could be related to the hex edits though, since those only affect open/save dialogs, and this doesn't look like one of those.

Also, what is it actually supposed to look like? To me it looks like just DPI scaling issues, which should be fixable by DpiAwareness=unaware.

**Comment by Spatzendame** [comment]
Here is savegame. I will show you how it's supposed to look like.
[QUICKSAVE.zip](https://github.com/user-attachments/files/18053804/QUICKSAVE.zip)

This is how the trading screen should look like. But you are right! This issue is not caused by the hex edits. It must be a general issue in your ddraw.dll or maybe I can try to resolve it using different GDI scaling options.
![Indy_trading screen](https://github.com/user-attachments/assets/edb3e7e6-49f3-4b44-a089-e5c7ec0e3003)

**Comment by Spatzendame** [comment]
Oh no, when I disable GDI scaling for the Indy3D.exe, the game no longer start and showing this error.
![Indy3d_novideo](https://github.com/user-attachments/assets/d9aa8c3d-c181-4f6f-baa5-9904f28dda2c)

**Comment by narzoul** [comment]
The game works fine for me with those setting. What are your in game display settings, your monitor resolution and your Windows DPI scale factor?

**Comment by Spatzendame** [comment]
I have to say I'm using Widescreen FOV Patcher from here:
https://www.gamepressure.com/download/indiana-jones-and-the-infernal-machine-widescreen-fov-fix-v10-mo/zb1503b

I can try whether that cause the isse but I don't believe.
My in game resolution is 3840x2160.
My Windows DPI scale factor is 200 %.

**Comment by narzoul** [comment]
It seems to be a resolution issue. If you use System scaling then the game can't run at the same resolution as your desktop.

I suggest you drop the in-game resolution to 1920x1080 and use ResolutionScale=display to scale it back up to 4k.

**Comment by Spatzendame** [comment]
Hey I now found out that ddraw.dll version 0.3.1 (for Win11 24H2) works instead of your current build 0.5.4.

Hmm when I use in game 1920x1080 I will have 4K resolution then?

**Comment by narzoul** [comment]
> Hmm when I use in game 1920x1080 I will have 4K resolution then?

Yes, if you use the suggested [ResolutionScale](https://github.com/narzoul/DDrawCompat/wiki/Configuration#resolutionscale) setting. If you're not sure how to do that, it's easiest if you press Shift+F11 in-game, change ResolutionScale to "display" from the dropdown list, then click on the Export button to save settings. (The setting is applied immediately even without export, but only until you exit the game).

**Comment by Spatzendame** [comment]
Tested version 0.5.2 (Win11 24H2 update) and it works too. Maybe 0.5.4 broke something or isn't fully up-to-date? Maybe it lacks some later fixes for Win11 24H2?

Thanks for the information which I will also try now.

**Comment by Spatzendame** [comment]
I tried in game 1920x1080 resolution without scaling to 4K just native 1080p and the same issue.
I think I will stay at ddraw.dll version 0.5.2 until your latest build will get the Win11 24H2 fix. I think that's the issue because all other builds from October works.

**Comment by narzoul** [comment]
> I tried in game 1920x1080 resolution without scaling to 4K just native 1080p and the same issue.

Which issue exactly?

> I think I will stay at ddraw.dll version 0.5.2 until your latest build will get the Win11 24H2 fix. I think that's the issue because all other builds from October works.

v0.5.4 already contains the fix that the older builds were updated with. It has nothing to do with any of these issues anyway.

**Comment by Spatzendame** [comment]
I forgot to set the GDI scaling unaware to "System (Advanced)" and now it works with 1080p and 0.5.4.
You are right, this issue is related to the GDI scaling. But for some reason previous builds can handle with it much better in 4K without the GDI scaling unaware. Something must be worse in 0.5.4 compared to 0.5.2.

I will now try the ResolutionScale setting to scale up to 4K. Otherwise I will stay at 0.5.2.

**Comment by IllustriousProsperus** [comment]
> I forgot to set the GDI scaling unaware to "System (Advanced)" and now it works with 1080p and 0.5.4. You are right, this issue is related to the GDI scaling. But for some reason previous builds can handle with it much better in 4K without the GDI scaling unaware. Something must be worse in 0.5.4 compared to 0.5.2.
> 
> I will now try the ResolutionScale setting to scale up to 4K. Otherwise I will stay at 0.5.2.

have you find a solution for the "invisible" save/load screen?

**Comment by narzoul** [comment]
> have you find a solution for the "invisible" save/load screen?

It's linked in my first comment.

**Comment by Spatzendame** [comment]
Yes, Narzoul has found a solution and I will make a full game patch that fixes all issues thanks to the help from Narzoul and others.

I also tried the Shift+F11 option menu and tweaked some settings. It really improves the game now!! The window boxes are also bigger in 1080p native using 4K upscaler and the game will have less issues on lower native resolutions. Thanks, I am so happy now!!!

**Comment by IllustriousProsperus** [comment]
> Yes, Narzoul has found a solution and I will make a full game patch that fixes all issues thanks to the help from Narzoul and others.
> 
> I also tried the Shift+F11 option menu and tweaked some settings. It really improves the game now!! The window boxes are also bigger in 1080p native using 4K upscaler and the game will have less issues on lower native resolutions. Thanks, I am so happy now!!!

I read that you use "dinput.dll", I don't use it and I have no changes or benefits, instead I'm using "dsoal-aldrv.dll" and "dsound.dll" ver 5.3.1.904 for 3D audio. The microstruttering needs to be resolved when indy moves in circles (with old versions of dgv2 the fps you had to be set to 59fps but now it's useless...). Oh I personally keep the resolution to minimum with the internal resolution upscaled.

**Comment by Spatzendame** [comment]
Yes, dgVoodoo2 had micro stuttering when using ddraw.dll from dgVoodoo2. It's no longer recommended to use dgVoodoo2 for this game, as great as it is! But who knows... maybe Dégé also manages to improve his ddraw.dll. I asked him about that. Always better to have both working.
The dinput.dll is required since Win11 24H2 for me.
DSOAL is a 3D sound wrapper.

**Comment by IllustriousProsperus** [comment]
> Yes, dgVoodoo2 had micro stuttering when using ddraw.dll from dgVoodoo2. It's no longer recommended to use dgVoodoo2 for this game, as great as it is! But who knows... maybe Dégé also manages to improve his ddraw.dll. I asked him about that. Always better to have both working. The dinput.dll is required since Win11 24H2 for me. DSOAL is a 3D sound wrapper.

With the new version of dgv2 (2.84) and a clean settings now the struttering is gone

**Comment by Spatzendame** [comment]
> > With the new version of dgv2 (2.84) and a clean settings now the struttering is gone

Maybe, but the rest of the issues like invisible save game list, window box crashes etc.?
For this reason the ddraw.dll by Narzoul is the best and makes dgVoodoo2 obsolete.

**Comment by smlu** [comment]
Just for info, OpenJones3D (https://github.com/smlu/OpenJones3D/releases/tag/v0.3.0) fixes this GDI dialog rendering issue.

**Comment by Spatzendame** [comment]
Hey smlu,
thank you very much for this information and your fantastic work on the game!!
I will try it soon!

But I have a question: why is it required to use Indy3D.exe from the CD v1.0 ?
What if I have CD v1.1 ?
And why not use the more updated GOG exe?

**Comment by smlu** [comment]
> But I have a question: why is it required to use Indy3D.exe from the CD v1.0 ?

RE was done on version v1.0 and since the OpenJones3D engine is not fully implemented yet it requires v1.0 exe to function properly (original functions are invoked in exe when not implemented in oj3d).  To use any other exe version, it would be necessary to partially RE and map the addresses of functions and global variables of the specific exe, which, as you can imagine, would be just too much work for a project like this.

**Comment by Spatzendame** [comment]
Hmm, why not doing a RE on latest official version 1.2 exe?
I also have added some fixes into the exe. For example a crash fix from here:
https://indiana-jones-forum.de/popup.php?action=showpost&id=115809

Does your project also fixes this?

And what about dgVoodoo2 ?? I assume it is no longer needed or get in conflict with your project?

**Comment by smlu** [comment]
> Hmm, why not doing a RE on latest official version 1.2 exe?

It's a many years long project, it started before game was released on GOG/Steam platform.
I don't have time to do another RE project, but I'd be willing to accept any suitable PR that brings support for v1.2.

> In my own patch project I also have added some fixes into the exe. For example a crash fix from here: https://indiana-jones-forum.de/popup.php?action=showpost&id=115809
> 
> Does your project also fixes this?

Yes, the 2-hand-held weapon fire bug is fixed in OpenJones3D

> And what about dgVoodoo2 ?? I assume it is no longer needed or get in conflict with your project?

Atm project is reconstructed 1:1 to match function parity with the original engine (with exception to bug fixes and engine enhancement). Thus, it still uses old DirectX 6 as the original, so any dx wrapper would be still required to run the game in cases your graphic drivers doesn't fully support DirectX 6.  Long term plan is to move away from dx to OpenGL.

**Comment by Spatzendame** [comment]
But the 1.2 version was released in early 2000.
So the RE doesn't even base on 1.1 exe?
You really need the very first unpatched 1.0 exe from the game? This isn't even easy to find on CD. Most CDs already have the 1.1 exe.
Okay all this doesn't matter if your project fixes everything what the official patches from LucasArts do - and even a lot more. This means you fixed also the Levels in the COG files? Also the Level 7 temple intro cutscene crash fix?

I think having DirectX 6 plus dgVoodoo2 is a very good setup. OpenGL is not supported by dgVoodoo2 so this would mean you wouldn't be able to have the benefits from dgVoodoo2 anymore. dgVoodoo2 adds a lot useful settings.

**Comment by smlu** [comment]
> You really need the very first unpatched 1.0 exe from the game?

Yes

> This means you fixed also the Levels in the COG files? Also the Level 7 temple intro cutscene crash fix?

Fixing bugs and errors in the scripts is outside the scope of the OpenJones3D project.
Fixed intro scripts can be found at: https://www.moddb.com/games/indiana-jones-and-the-infernal-machine/downloads/cog-level-start-fix-patch-12

> OpenGL is not supported by dgVoodoo2 so this would mean you wouldn't be able to have the benefits from dgVoodoo2 anymore. dgVoodoo2 adds a lot useful settings.

The idea is to use OpenGL and GLSL shaders to enhance game experience, e.g.: add dynamic lighting, shadows etc...
Besides that, project aims to eventually port engine to other platforms (Linux, macOS etc...) where DirectX is not supported.

**Comment by smlu** [comment]
> Hmm then can we get in contact via email about 1.0 exe?

There is discord server where you can find all infos about v1.0: https://discord.gg/W7EV2nhKgh

> Okay this are the original scripts from the 1.2 update. If you mix files from 1.2 with the exe from 1.0 is not recommended. Or di they have dedicated functions? Can you mix scripts with every exe?

There wasn't any fundamental change made to COG system and the scripts between v1.2 and v1.0 are compatible.

> Why not Vulkan instead of OpenGL?

Vulkan would be an overkill for the game engine like this. 

> With the source code re-assembled someone could remake the game in Unreal Engine 5?

Since the code is in C it should be compatible with UE 5 codebase, but would probably make more sense to just create new engine than relay on the original code.

**Comment by Spatzendame** [comment]
Can I simply install the GOG version and then replace the Indy3D.exe with the v1.0 from the CD?
Maybe you can accept the Indy3D.exe's from v1.1, v1.2, GOG and Steam aswell without making a new RE?

**Comment by smlu** [comment]
Yes you can any distribution and then just replace the exe.


> Maybe you can accept the Indy3D.exe's from v1.1, v1.2, GOG and Steam aswell without making a new RE?

Each exe has different address offsets in the binary that's why it only works with v1.0. 
REing other exes would just  take too much time to collect all required  addresses for functions and global variables.
There are some binary comparison tools that can help map addresses for functions (e.g.: bindiff), but in my experience it still requires a lot of manual input to find and correct all false matches and unmatched addresses + global var addresses.
Tho, I'll except any PR if someone could map v1.2 to v1.0.

**Comment by Spatzendame** [comment]
Thank you. I bought a CD with v1.0 because my actual CD is a re-release already containing the v1.1 version.
Interesting AI upscale project. I also wanted to do that. I hope it was a very good AI adding new details on textures.

**Comment by Spatzendame** [comment]
Thank you. Can you show me the CD containing the 1.0 exe?
Maybe I have to buy the US version. I tried all german CD releases and never got the 1.0 exe.
Maybe the game launched later in the EU and got already the first patch?

**Comment by Spatzendame** [comment]
I bought the US version. I want a proper 1.0 exe. Not a downgrade.
Already found the 1.0 exe on archive.org.

**Comment by smlu** [comment]
> I bought the US version. I want a proper 1.0 exe. Not a downgrade.
> Already found the 1.0 exe on archive.org.

Does it work now?

> You are sure when have a 1.1 Installation or even 1.2 and replace the exe with the 1.0 version it won't get in conflict with the .GOB archives where the game files are packed? Maybe there are changes that require the updated exe? Ah wait the 1.1 patch was only some megabyte or kilobyte so it won't change the .GOB archives. At least adding a Cog folder with updated level files which may be already included in the archives on 1.1 master disks.

The exe is not tied to game assets so nothing breaks when using different resources.
You might also want to use these COG scripts as they fix what wasn't fixed in v1.2 patch: https://www.moddb.com/games/indiana-jones-and-the-infernal-machine/downloads/cog-level-start-fix-patch-12

**Comment by Spatzendame** [comment]
Does it only affect the italian version? Normally the language doesn't have an affect to the Cog files. Also the patch 1.2 is the same for all languages. 
In my experience only the first level glitched right after the "Indiana Jones" logo and the second was a crash when entering level 7 temple of palawan. I fixed both.

**Comment by Spatzendame** [comment]
I will report every issue I encounter.

I noticed the Steam version scripts already have some changes included compared to original 1.2 scripts. The GOG version doesn't have any Cog folder with scripts. GOG said that it's a 1.2 master version where the scripts are packed in the main archives CD1.GOB & CD2.GOB but I have checked it and it's not the case.
